lib/http.c

```diff
@@ -4270,6 +4270,7 @@ CURLcode Curl_http_readwrite_headers(struct Curl_easy *data,
                 data->state.disableexpect = TRUE;
                 DEBUGASSERT(!data->req.newurl);
                 data->req.newurl = strdup(data->state.url);
+                streamclose(conn, "Stop sending data before everything sent");
```

Comment:
the info line for this block says it's handling a 417 while waiting for the 100, but your issue says that isn't what's happening. instead the time passed and data was partially sent. so is it correct to put it in this section specifically. like wouldn't data->req.writebytecount be nonzero for what you describe

Reply:
It is still "hoping" for a 100, but it's past the point of sitting there idle waiting. It seems to handle the header (whether 100 or 417) as soon as it's received, so this looks to me like the right spot to put it. If it's not 417 the connection is closed in the same way in the block a few lines afterward. Yes, writebytecount is nonzero at this point in this failure scenario.

Reply:
what i mean is why not put some conditionals on it. isn't that info line misleading? aren't they mutually exclusive so if we receive 417 reply then we aren't going to receive 100 reply

~~~diff
diff --git a/lib/http.c b/lib/http.c
index 860843a..fd63dc0 100644
--- a/lib/http.c
+++ b/lib/http.c
@@ -4266,11 +4266,17 @@ CURLcode Curl_http_readwrite_headers(struct Curl_easy *data,
               if((k->httpcode == 417) && data->state.expect100header) {
                 /* 417 Expectation Failed - try again without the Expect
                    header */
-                infof(data, "Got 417 while waiting for a 100");
+                if(!k->writebytecount &&
+                   data->req.exp100 == EXP100_AWAITING_CONTINUE) {
+                  infof(data, "Got HTTP failure 417 while waiting for a 100");
+                }
+                else {
+                  infof(data, "Got HTTP failure 417 while sending data");
+                  streamclose(conn, "Stop sending data before everything sent");
+                }
                 data->state.disableexpect = TRUE;
                 DEBUGASSERT(!data->req.newurl);
                 data->req.newurl = strdup(data->state.url);
-                streamclose(conn, "Stop sending data before everything sent");
                 Curl_done_sending(data, k);
               }
               else if(data->set.http_keep_sending_on_error) {
~~~